### PR TITLE
Origin/adding lapack2flash support

### PR DESCRIPTION
**Pull request by ItsPitt**
Part 1 of making LAPACK2FLASH. Made a new directory and added it to the libflame build process. All .c files in lapack2flash/ are currently just copies of the lapack2flamec/ files.

### CODE REVIEW

File: configure [hunk 1]
```diff
@@ -1514,6 +1516,7 @@ Optional Features:
                           if this option is enabled, then
                           external-lapack-for-subproblems MUST be disabled.
                           (Disabled by default.)
+  --enable-lapack2flash   (In test at the moment)
```
Comment (iotamudelta) [inline configure:1519]: Please add comment here.

File: src/map/lapack2flash/README [hunk 2]
```diff
@@ -0,0 +1,58 @@
+
+lapack2flash
+============
+
+
+TODO
```
Comment (iotamudelta) [inline src/map/lapack2flash/README:6]: Please do. :-)

File: src/map/lapack2flash/README [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+Directory structure
+-------------------
+
+lapack2flame
```
Comment (iotamudelta) [inline src/map/lapack2flash/README:11]: flame -> flash

File: src/map/lapack2flash/README [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  - check               :: includes LAPACK arguments check routines.
+  - f2c
```
Comment (iotamudelta) [inline src/map/lapack2flash/README]: Comment that this and check are softlinks to the lapack2flame versions

File: src/map/lapack2flash/README [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+                           FLA_*.c will replace those files.
+  + FLA_*.c             :: LAPACK interfaces that redirect functionality to libflame.
+
+
+How to use
+----------
+
+configuration:
+
+TODO
```
Comment (iotamudelta) [inline src/map/lapack2flash/README:23]: Same.

File: src/map/lapack2flash/README [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+./configure \
+            --enable-lapack2flash \
+            TODO
```
Comment (iotamudelta) [inline src/map/lapack2flash/README:27]: Same

File: src/map/lapack2flash/README [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+LAPACK test suite
+-----------------
+
+TODO
```
Comment (iotamudelta) [inline src/map/lapack2flash/README:32]: Same

File: src/map/lapack2flash/README [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+
+F2C scripts
+-----------------
+
+TODO
```
Comment (iotamudelta) [inline src/map/lapack2flash/README:38]: Same

**Review by iotamudelta — COMMENTED** [review]
Missing some documentation. Also please add LaTeX documentation into the `docs` documentation.

File: src/map/lapack2flash/FLASH_potrf.c [hunk 9]
```diff
@@ -41,16 +41,25 @@
   FLA_Obj      A;                                               \
   FLA_Error    e_val;                                           \
   FLA_Error    init_result;                                     \
+  dim_t        depth = DEPTH_SIZE;                              \
```
Comment (ItsPitt) [inline src/map/lapack2flash/FLASH_potrf.c]: (unsigned long)

File: src/map/lapack2flash/FLASH_gelsd.c [hunk 10]
```diff
@@ -51,17 +51,17 @@
   At this moment, this routine is redirected to GELSS.
 */
```
Comment (ItsPitt) [inline src/map/lapack2flash/FLASH_gelsd.c:53]: 2 of these files hard no real changes. And no clear way to use FLASH

File: src/map/lapack2flash/README [hunk 11]
```diff
@@ -9,25 +9,9 @@ Directory structure
...
-                           these files are likely to be never changed even if LAPACK evolves.
-      - util            :: necessary f2c sources files; some of them are modified from original ones.
-    - netlib
-      - 3.5.0.tar       :: tar ball includes source files from LAPACK 3.5.0.
-      - support.list    :: a list of files that will not be f2c'ed. 
-                           FLA_*.c will replace those files.
+  - check               :: Symbolic link to the directory in lapack2flamec/
+  - f2c                 :: Symbolic link to the directory in lapack2flamec/
+
   + FLA_*.c             :: LAPACK interfaces that redirect functionality to libflame.
```
Comment (ItsPitt) [inline src/map/lapack2flash/README:15]: Still need to add a lot

File: src/map/lapack2flash/FLASH_lapack2flash_util_defs.h [hunk 12]
```diff
@@ -86,7 +86,7 @@ extern int FLAME_invert_ztau( FLA_Obj t );
 extern int FLAME_QR_piv_preorder( FLA_Obj A, int *jpiv_lapack, int *jpiv_fla );
 
 // TODO, Let a user change these at run time
-static dim_t FLASH_depth          = 1;
```
Comment (ItsPitt) [inline src/map/lapack2flash/FLASH_lapack2flash_util_defs.h]: This was wrong and undone in a later commit

File: src/lapack/dec/q/qrut/front/flamec/FLASH_QR_UT_create_T.c [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+  }
+  else // if ( FLA_Obj_col_stride( A ) == 1 )
+  {
+    rs_T = k;
+    cs_T = 1;
+  }
+
+  // Create a b_alg x k matrix to hold the block Householder transforms that
+  // will be accumulated within the QR factorization algorithm.
+  FLASH_Obj_create( datatype, b_alg, k, rs_T, cs_T, T );
```
Comment (ItsPitt) [inline src/lapack/dec/q/qrut/front/flamec/FLASH_QR_UT_create_T.c]: I'm assuming this is correct. but not 100% at this point. Not sure if I need to use my user set depth and block size

File: src/lapack/dec/q/qrut/front/flamec/FLASH_QR_UT_form_Q.c [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+
+        // Set Q identify
+        FLA_Set_to_identity( Q );
+
+        // Q = H_{0} H_{1} ... H_{k-1}
+        FLASH_Apply_Q_UT_create_workspace_side( FLA_LEFT, T, Q, &W );
+        r_val = FLA_Apply_Q_UT( FLA_LEFT, FLA_NO_TRANSPOSE,
+                                FLA_FORWARD, FLA_COLUMNWISE,
+                                A, T, W, Q );
+        //FLASH_Obj_free( &W );  Need to see if it's alright to sub these out, this might cause a mem leak
```
Comment (ItsPitt) [inline src/lapack/dec/q/qrut/front/flamec/FLASH_QR_UT_form_Q.c]: Had to switch this to _without_buffer. Not sure if this is correct, but I was getting a undefined mem reference when trying to free W with this one.

File: src/lapack/red/bidiagut/front/flamec/FLASH_Bidiag_UT_create_T.c [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+  {
+    rs_T = k;
+    cs_T = 1;
+  }
+
+  // Create two b_alg x k matrices to hold the block Householder transforms
+  // that will be accumulated within the bidiagonal reduction algorithm.
+  // If the matrix dimension has a zero dimension, apply_q complains it.
+  //if ( TU != NULL ) FLASH_Obj_create( datatype, b_alg, k, rs_T, cs_T, TU );
+  //if ( TV != NULL ) FLASH_Obj_create( datatype, b_alg, k, rs_T, cs_T, TV );
```
Comment (ItsPitt) [inline src/lapack/red/bidiagut/front/flamec/FLASH_Bidiag_UT_create_T.c:52]: Unlike FLASH_QR_UT_create_T, this was causing mismatched obj types I believe. I need to figure out which way is correct

File: src/lapack/red/bidiagut/front/flamec/FLASH_Bidiag_UT_form_U_ext.c [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+                FLA_Part_2x2( A,    &ATL, &ATR,
+                                    &ABL, &ABR,    1, b+1, FLA_TL );
+
+                FLASH_Obj_create( FLA_INT, b+1, 1, depth, &blocksize, &p );
+                FLA_Part_2x1( p,    &pt,
+                                    &pb, 1, FLA_BOTTOM );
+                FLASH_Set( FLA_ONE, pt );
+                FLASH_Set( FLA_ZERO, pb );
+                FLA_Apply_pivots ( FLA_RIGHT, FLA_NO_TRANSPOSE, p, ABL );
+                //FLASH_Apply_pivots ( FLA_RIGHT, FLA_NO_TRANSPOSE, p, ABL );
```
Comment (ItsPitt) [inline src/lapack/red/bidiagut/front/flamec/FLASH_Bidiag_UT_form_U_ext.c:62]: This resulted in a "Function or conditional branch/case not yet implemented" error. This should be changed once it is

File: src/lapack/util/app/qut/front/flamec/FLASH_Apply_Q_UT_create_workspace.c [hunk 17]
```diff
@@ -67,3 +67,64 @@ FLA_Error FLASH_Apply_Q_UT_create_workspace( FLA_Obj TW, FLA_Obj B, FLA_Obj* W )
...
+
+	// The scalar length of W should be the algorithmc/storage blocksize
+	// encoded in TW.
+	m = b_alg;
+
+	// Query the scalar (not element) width of the right-hand side
+	// matrix B.
+	//n = FLASH_Obj_scalar_width( B );
+
+	if      ( side == FLA_LEFT  ) n = FLASH_Obj_scalar_width( B );
```
Comment (ItsPitt) [inline src/lapack/util/app/qut/front/flamec/FLASH_Apply_Q_UT_create_workspace.c:120]: I'll fix this spacing issue

File: src/map/lapack2flash/FLASH_orgbr.c [hunk 18]
```diff
@@ -183,13 +203,26 @@
...
+                                  T,                                    \
+                                  FLA_CONJ_TRANSPOSE,                   \
+                                  A,                                    \
+                                  FLASH_depth,                          \
+                                  &FLASH_blocksize );                   \
     }                                                                   \
   }                                                                     \
                                                                         \
-  FLASH_Obj_free( &T );                                                 \
+  FLASH_Obj_free_without_buffer( &T );                                  \
```
Comment (ItsPitt) [inline src/map/lapack2flash/FLASH_orgbr.c:226]: See above

File: src/map/lapack2flash/FLASH_orgbr.c [hunk 19]
```diff
@@ -100,11 +100,12 @@
     FLA_Part_2x1( t, &t,                                                \
                      &T, FLA_Obj_min_dim( Ah ), FLA_TOP );              \
                                                                         \
+/* TODO: I believe this should be adding a buffer and it is not*/       \
```
Comment (ItsPitt) [inline src/map/lapack2flash/FLASH_orgbr.c:103]: As noted, I believe this should be creating T with a buffer. However it is not and causing a few issues. I'll need to circle back to this

**Comment by iotamudelta** [comment]
Integrated upstream

**Pull request closed without merging**